File: src/main/java/exnihilocreatio/json/CustomCompostableJson.java
package exnihilocreatio.json;

import com.google.gson.*;
import exnihilocreatio.registries.types.Compostable;
import exnihilocreatio.texturing.Color;
import exnihilocreatio.util.BlockInfo;
import exnihilocreatio.util.StackInfo;

import java.lang.reflect.Type;

public class CustomCompostableJson implements JsonDeserializer<Compostable>, JsonSerializer<Compostable> {
    @Override
    public JsonElement serialize(Compostable src, Type typeOfSrc, JsonSerializationContext context) {
        JsonObject obj = new JsonObject();
        obj.addProperty("value", src.getValue());
        obj.add("compostBlock", context.serialize(src.getCompostBlock(), StackInfo.class));
        if (!src.getColor().equals(Color.INVALID_COLOR)){
            obj.add("color", context.serialize(src.getColor(), Color.class));
        }

        return obj;
    }

    @Override
    public Compostable deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        JsonHelper helper = new JsonHelper(json);
        JsonObject obj = json.getAsJsonObject();

        Float value = (float)helper.getDouble("value");
        Color color = Color.INVALID_COLOR;
        if (obj.has("color")) {
            color = context.deserialize(json.getAsJsonObject().get("color"), Color.class);
        }

        BlockInfo result = context.deserialize(json.getAsJsonObject().get("compostBlock"), BlockInfo.class);
        if (result.getMeta() == -1)
            result.setMeta(0);


        return new Compostable(value, color, result);
    }
}

File: src/main/java/exnihilocreatio/json/CustomCompostableJson.kt
package exnihilocreatio.json

import com.google.gson.*
import exnihilocreatio.registries.types.Compostable
import exnihilocreatio.texturing.Color
import exnihilocreatio.util.BlockInfo

import java.lang.reflect.Type

object CustomCompostableJson : JsonDeserializer<Compostable>, JsonSerializer<Compostable> {
    override fun serialize(src: Compostable, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val obj = JsonObject()
        obj.addProperty("value", src.value)
        obj.add("compostBlock", context.serialize(src.compostBlock, BlockInfo::class.java))
        if (src.color != Color.INVALID_COLOR) {
            obj.add("color", context.serialize(src.color, Color::class.java))
        }

        return obj
    }

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Compostable {
        val helper = JsonHelper(json)
        val obj = json.asJsonObject

        val value = helper.getDouble("value").toFloat()
        var color = Color.INVALID_COLOR
        if (obj.has("color")) {
            color = context.deserialize(json.asJsonObject.get("color"), Color::class.java)
        }

        val result = context.deserialize<BlockInfo>(json.asJsonObject.get("compostBlock"), BlockInfo::class.java)

        return Compostable(value, color, result)
    }
}
